### PR TITLE
Retry test_psu.py::TestPsuApi::test_power if power not within tolerance

### PR DESCRIPTION
**Pull request by smagarwal-arista**
### Description of PR

Summary:
If `test_psu.py::TestPsuApi::test_power`  fails on `abs(power - (voltage*current)) < power*0.1` then retry the test.  The test is repeated for a maximum of three times until it passes, else it is failed. This change is introduced to account for the stringent tolerance level of 10% and the errors that might be caused due to time differences in reading each of the parameters (current, voltage and power). 

Fixes # (issue)
Addresses issue: https://github.com/aristanetworks/sonic-qual.msft/issues/209 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [x] 202311
- [x] 202405

#### What is the motivation for this PR?
Improve the test to account for stringent tolerance level and asymmetric reading of parameters for power calculation. 
#### How did you do it?
If the test fails for the condition `abs(power - (voltage*current)) < power*0.1`, the test will collect fresh parameters (voltage, current and power) from the PSU telemetry and recheck the condition. It is limited to a maximum of 3 attempts. 
#### How did you verify/test it?
Tested internally by varying the tolerance level and verifying the number of api calls. 
#### Any platform specific information?
N/A
#### Supported testbed topology if it's a new test case?
N/A
### Documentation
N/A

### CODE REVIEW

File: tests/platform_tests/api/test_psu.py [hunk 1]
```diff
@@ -206,66 +207,75 @@ def test_power(self, duthosts, enum_rand_one_per_hwsku_hostname, localhost, plat
...
+                            self.expect(isinstance(low_threshold, float),
+                                        "PSU {} voltage low threshold appears incorrect".format(psu_id))
+                    if high_threshold is not None and low_threshold is not None:
+                        self.expect(voltage < high_threshold and voltage > low_threshold,
+                                    "Voltage {} of PSU {} is not in between {} and {}"
+                                    .format(voltage, psu_id, low_threshold, high_threshold))
+
+                break
+
         self.assert_expectations()
```
Comment (vvolam) [inline tests/platform_tests/api/test_psu.py:279]: How does the self.assert_expectations() evaluate the 3 test run values? What happens for different combinations of the expected values? For example:

Test1: True
Test2: False
Test3: True

In this combination, does the test raise assert or passes?

Comment (smagarwal-arista) [reply tests/platform_tests/api/test_psu.py:279]: The test would fail in this scenario. I have an update to address it, will commit the changes. Thanks.

Comment (vvolam) [reply tests/platform_tests/api/test_psu.py:283]: @smagarwal-arista With new changes, does the test pass in above scenario? Also, does the test exit on first Passed case?

Comment (smagarwal-arista) [reply tests/platform_tests/api/test_psu.py:283]: Yes, the test would exit the retry loop if all expects pass.  Retry is only targeted for expect failing on `abs(power - (voltage*current)) < power*0.1` condition. Any other expect failures would exit the retry loop for that PSU and eventually result in a failed test.

**Review by vvolam — COMMENTED** [review]
Looks good to me. 

@smagarwal-arista  As we are modifying this function already, can you modularize the function for better readability?

**Comment by smagarwal-arista** [comment]
> Looks good to me.
> 
> @smagarwal-arista As we are modifying this function already, can you modularize the function for better readability?

Sure, will update the function

**Review by vvolam — APPROVED** [review]
Thank you. Looks good to me.

**Comment by vvolam** [comment]
@prgeor could you review the PR?

**Review by prgeor — COMMENTED** [review]
@roy-sror @keboliu @Junchao-Mellanox can you review

**Review by prgeor — COMMENTED** [review]
@Junchao-Mellanox @keboliu @roy-sror can you review?

File: tests/platform_tests/api/test_psu.py [hunk 2]
```diff
@@ -204,68 +215,52 @@ def test_power(self, duthosts, enum_rand_one_per_hwsku_hostname, localhost, plat
...
+                name = psu.get_name(platform_api_conn, psu_id)
+                if name in self.psu_skip_list:
+                    logger.info("skipping check for {}".format(name))
+                else:
+                    voltage = self.get_psu_parameter(psu_info, "voltage", psu.get_voltage, "voltage")
+                    current = self.get_psu_parameter(psu_info, "current", psu.get_current, "current")
+                    power = self.get_psu_parameter(psu_info, "power", psu.get_power, "power")
+
+                    failure_occured = self.get_len_failed_expectations() > failure_count
+                    if current is not None and voltage is not None and power is not None:
```
Comment (JibinBao) [inline tests/platform_tests/api/test_psu.py:236]: 'if current is not None and voltage is not None and power is not None: ' is changed to  **if current  and voltage and power :**
Maybe It is more concise?

File: tests/platform_tests/api/test_psu.py [hunk 3]
```diff
@@ -204,68 +215,52 @@ def test_power(self, duthosts, enum_rand_one_per_hwsku_hostname, localhost, plat
...
+                    if self.expect(powergood_status is not None,
+                                   "Failed to retrieve operational status of PSU {}".format(psu_id)):
+                        self.expect(powergood_status is True, "PSU {} is not operational".format(psu_id))
+
+                    high_threshold = self.get_psu_parameter(psu_info, "voltage_high_threshold",
+                                                            psu.get_voltage_high_threshold, "high voltage threshold")
+                    low_threshold = self.get_psu_parameter(psu_info, "voltage_low_threshold",
+                                                           psu.get_voltage_low_threshold, "low voltage threshold")
+
+                    if high_threshold is not None and low_threshold is not None:
```
Comment (JibinBao) [inline tests/platform_tests/api/test_psu.py:257]: Same as line 236. if high_threshold and low_threshold:

**Review by JibinBao — COMMENTED** [review]
In the case, you want to retry the same test logic for 3 times. I think you don't need to desgin the retry mechanisam by yourself, you can use the exsiting function:
wait_until(timeout, interval, delay, condition, *args, **kwargs).
Using the function of wait_until will make the test more concise

**Comment by smagarwal-arista** [comment]
> In the case, you want to retry the same test logic for 3 times. I think you don't need to desgin the retry mechanisam by yourself, you can use the exsiting function: wait_until(timeout, interval, delay, condition, *args, **kwargs). Using the function of wait_until will make the test more concise

Thanks, I'll take a look

**Comment by StormLiangMS** [comment]
hi @smagarwal-arista could you address the comments?

**Comment by smagarwal-arista** [comment]
> In the case, you want to retry the same test logic for 3 times. I think you don't need to desgin the retry mechanisam by yourself, you can use the exsiting function: wait_until(timeout, interval, delay, condition, *args, **kwargs). Using the function of wait_until will make the test more concise

Resolved in the latest commit. Now using wait_until function.

File: tests/platform_tests/api/test_psu.py (hunk 2 above)
Comment (smagarwal-arista) [reply tests/platform_tests/api/test_psu.py]: Resolved in the latest commit.

File: tests/platform_tests/api/test_psu.py (hunk 3 above)
Comment (smagarwal-arista) [reply tests/platform_tests/api/test_psu.py]: Resolved in the latest commit.

**Review by prgeor — APPROVED** [review]

**Comment by bingwang-ms** [comment]
MSADO: 28386639

**Pull request merged**